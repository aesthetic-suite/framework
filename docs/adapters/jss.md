# JSS

Provides [JSS](https://github.com/cssinjs/jss) support.

```javascript
import Aesthetic from 'aesthetic';
import JSSAdapter from 'aesthetic-adapter-jss';

const aesthetic = new Aesthetic(new JSSAdapter());
```

## Unified Syntax

To make use of the unified syntax, simply import the adapter from the unified path.

```javascript
import JSSAdapter from 'aesthetic-adapter-jss/unified';
```

## Plugins

[JSS plugins](https://github.com/cssinjs/jss/blob/master/docs/plugins.md) can be customized by
instantiating a new JSS instance and passing it the adapter.

```javascript
import Aesthetic from 'aesthetic';
import JSSAdapter from 'aesthetic-adapter-jss';
import { create } from 'jss';
import preset from 'jss-preset-default';

const jss = create(preset());
const aesthetic = new Aesthetic(new JSSAdapter(jss));
```

## Configuration

Furthermore, options to be passed to `jss.createStyleSheet` can be defined through the 2nd argument
of the constructor.

```javascript
import Aesthetic from 'aesthetic';
import JSSAdapter from 'aesthetic-adapter-jss';

const aesthetic = new Aesthetic(new JSSAdapter(null, {
  media: 'screen',
  // ...
});
```
