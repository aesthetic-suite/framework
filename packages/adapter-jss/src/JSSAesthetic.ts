import Aesthetic, {
  AestheticOptions,
  ClassName,
  Ruleset,
  Sheet,
  StyleName,
  SheetMap,
} from 'aesthetic';
import { JSS, StyleSheet as JSSSheet } from 'jss';
import { NativeBlock, ParsedBlock } from './types';

export default class JSSAesthetic<Theme extends object> extends Aesthetic<
  Theme,
  NativeBlock,
  ParsedBlock
> {
  jss: JSS;

  keyframes: { [animationName: string]: string } = {};

  sheets: { [styleName: string]: JSSSheet<any> } = {};

  constructor(jss: JSS, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.jss = jss;

    this.syntax
      .on('attribute', this.handleNested)
      .on('charset', this.handleCharset)
      .on('css', this.handleCss)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('import', this.handleImport)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('support', this.handleSupport)
      .on('viewport', this.handleViewport);
  }

  flushStyles(styleName: StyleName) {
    const sheet = this.sheets[styleName];

    if (sheet) {
      sheet.attach();
    }
  }

  isParsedBlock(block: ParsedBlock): boolean {
    return typeof block === 'string';
  }

  protected processStyleSheet(
    styleSheet: SheetMap<NativeBlock>,
    styleName: StyleName,
  ): SheetMap<ParsedBlock> {
    this.sheets[styleName] = this.jss.createStyleSheet<any>(styleSheet, {
      media: 'screen',
      meta: styleName,
    });

    return this.sheets[styleName].classes;
  }

  protected transformToClassName(styles: ParsedBlock[]): ClassName {
    return styles.join(' ');
  }

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleCharset = (sheet: Sheet<NativeBlock>, charset: string) => {
    sheet.addAtRule('@charset', charset);
  };

  private handleCss = (css: string) => {
    this.getStyleSheetManager().injectStatements(css);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#fallbacks
  private handleFallback = (
    ruleset: Ruleset<NativeBlock>,
    name: keyof NativeBlock,
    value: any[],
  ) => {
    const fallbacks = value.map(fallback => ({ [name]: fallback }));
    const current = ruleset.properties.fallbacks || [];

    ruleset.addProperty('fallbacks', [...current, ...fallbacks]);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#font-face
  private handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: Ruleset<NativeBlock>[]) => {
    const current = (sheet.atRules['@font-face'] as Ruleset<NativeBlock>[]) || [];

    sheet.addAtRule('@font-face', [...current, ...fontFaces]);
  };

  // https://github.com/cssinjs/jss-global
  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    const current = (sheet.atRules['@global'] as Sheet<NativeBlock>) || sheet.createSheet();

    current.addRuleset(ruleset);

    sheet.addAtRule('@global', current);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleImport = (sheet: Sheet<NativeBlock>, paths: string[]) => {
    sheet.addAtRule('@import', paths);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#keyframes-animation
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframe: Ruleset<NativeBlock>,
    animationName: string,
  ) => {
    let name = this.keyframes[animationName];

    // Used to avoid global collision
    if (!name) {
      name = `${animationName}-${Object.keys(this.keyframes).length}`;

      this.keyframes[animationName] = name;
    }

    sheet.addAtRule(`@keyframes ${name}`, keyframe);
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@media ${query}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`&${selector}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, this.syntax.injectKeyframes(value, this.keyframes).join(', '));
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleSupport = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@supports ${query}`, value);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleViewport = (sheet: Sheet<NativeBlock>, ruleset: Ruleset<NativeBlock>) => {
    sheet.addAtRule('@viewport', ruleset);
  };
}
