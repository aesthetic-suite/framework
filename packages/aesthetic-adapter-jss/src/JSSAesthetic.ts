/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Aesthetic, {
  AestheticOptions,
  ClassName,
  Keyframes,
  Ruleset,
  Sheet,
  StyleName,
  StyleSheetMap,
} from 'aesthetic';
import { JSS, StyleSheet as JSSSheet } from 'jss';
import { NativeBlock, ParsedBlock } from './types';

export default class JSSAdapter<Theme> extends Aesthetic<Theme, NativeBlock, ParsedBlock> {
  jss: JSS;

  sheets: { [styleName: string]: JSSSheet<any> } = {};

  constructor(jss: JSS, options: Partial<AestheticOptions> = {}) {
    super(options);

    this.jss = jss;

    this.syntax
      .on('attribute', this.handleNested)
      .on('charset', this.handleCharset)
      .on('fallback', this.handleFallback)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('import', this.handleImport)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested)
      .on('support', this.handleSupport)
      .on('viewport', this.handleViewport);
  }

  protected processStyleSheet(
    styleSheet: object,
    styleName: StyleName,
  ): StyleSheetMap<ParsedBlock> {
    this.sheets[styleName] = this.jss
      .createStyleSheet<any>(styleSheet, {
        media: 'screen',
        meta: styleName,
        classNamePrefix: `${styleName}-`,
      })
      .attach();

    return this.sheets[styleName].classes;
  }

  protected transformToClassName(styles: (NativeBlock | ParsedBlock)[]): ClassName {
    const legitStyles: ParsedBlock[] = [];
    const tempStylesheet: { [key: string]: NativeBlock } = {};
    let counter = 0;

    styles.forEach(style => {
      if (typeof style === 'string') {
        legitStyles.push(style);
      } else if (typeof style === 'object' && style !== null) {
        tempStylesheet[`inline-${counter}`] = style;
        counter += 1;
      }
    });

    if (counter > 0) {
      legitStyles.push(...Object.values(this.processStyleSheet(tempStylesheet, 'inline-dynamic')));
    }

    return legitStyles.join(' ');
  }

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleCharset = (sheet: Sheet<NativeBlock>, charset: string) => {
    sheet.addAtRule('@charset', charset);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#fallbacks
  private handleFallback = (
    ruleset: Ruleset<NativeBlock>,
    name: keyof NativeBlock,
    value: any[],
  ) => {
    const fallbacks = value.map(fallback => ({ [name]: fallback }));
    const current = ruleset.properties.fallbacks || [];

    ruleset.addProperty('fallbacks', [...current, ...fallbacks]);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#font-face
  private handleFontFace = (sheet: Sheet<NativeBlock>, fontFaces: NativeBlock[]) => {
    const current = (sheet.atRules['@font-face'] as NativeBlock[]) || [];

    sheet.addAtRule('@font-face', [...current, ...fontFaces]);
  };

  // https://github.com/cssinjs/jss-global
  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    const current = (sheet.atRules['@global'] as Sheet<NativeBlock>) || new Sheet<NativeBlock>();

    current.addRuleset(ruleset);

    sheet.addAtRule('@global', current);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleImport = (sheet: Sheet<NativeBlock>, paths: string[]) => {
    sheet.addAtRule('@import', paths);
  };

  // https://github.com/cssinjs/jss/blob/master/docs/json-api.md#keyframes-animation
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframes: Keyframes<NativeBlock>,
    animationName: string,
  ) => {
    sheet.addAtRule(`@keyframes ${animationName}`, keyframes);
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@media ${query}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`&${selector}`, value);
  };

  // https://github.com/cssinjs/jss-nested#use--to-reference-selector-of-the-parent-rule
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, `$${value}`);
    } else {
      ruleset.addProperty(name, value);
    }
  };

  // https://github.com/cssinjs/jss-nested#use-at-rules-inside-of-regular-rules
  private handleSupport = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@supports ${query}`, value);
  };

  // https://github.com/cssinjs/jss/blob/master/packages/jss/tests/integration/sheet.js#L144
  private handleViewport = (sheet: Sheet<NativeBlock>, ruleset: Ruleset<NativeBlock>) => {
    sheet.addAtRule('@viewport', ruleset);
  };
}
