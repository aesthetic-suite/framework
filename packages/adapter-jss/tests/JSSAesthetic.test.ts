/* eslint-disable jest/expect-expect */

import { create } from 'jss';
// @ts-ignore
import preset from 'jss-preset-default';
import {
  cleanupStyleElements,
  getFlushedStyles,
  renderAndExpect,
  DIRECTIONS,
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_UNIFIED_LOCAL_FULL,
  SYNTAX_ATTRIBUTE,
  SYNTAX_DESCENDANT,
  SYNTAX_FALLBACKS,
  SYNTAX_FONT_FACE,
  SYNTAX_FONT_FACE_MIXED,
  SYNTAX_FONT_FACE_MULTIPLE,
  SYNTAX_GLOBAL,
  SYNTAX_KEYFRAMES,
  SYNTAX_MEDIA_QUERY,
  SYNTAX_MULTI_SELECTOR,
  SYNTAX_PROPERTIES,
  SYNTAX_PSEUDO,
  SYNTAX_SUPPORTS,
  SYNTAX_VIEWPORT,
  SYNTAX_CHARSET,
  SYNTAX_IMPORT,
  SYNTAX_IMPORT_MULTIPLE,
  SYNTAX_MEDIA_QUERY_NESTED,
  SYNTAX_KEYFRAMES_INLINE,
  KEYFRAME_SLIDE_PERCENT,
  SYNTAX_FONT_FACES_INLINE,
  SYNTAX_RAW_CSS,
} from 'aesthetic/lib/testUtils';
import JSSAesthetic from '../src/JSSAesthetic';

jest.mock('uuid/v4', () => () => 'uuid');

describe('JSSAesthetic', () => {
  let instance: JSSAesthetic<any>;

  beforeEach(() => {
    const jss = create();
    jss.setup(preset());

    instance = new JSSAesthetic(jss);
  });

  afterEach(() => {
    cleanupStyleElements();
  });

  DIRECTIONS.forEach(dir => {
    // eslint-disable-next-line jest/valid-describe
    describe(dir.toUpperCase(), () => {
      it('converts and transforms inline styles', () => {
        expect(instance.transformStyles(['foo', { margin: 0 }, { padding: 2 }], { dir })).toBe(
          dir === 'ltr'
            ? 'foo inline-0-0-1-1 inline-1-0-1-2'
            : 'foo inline-0-0-25-1 inline-1-0-25-2',
        );
        expect(getFlushedStyles()).toMatchSnapshot();

        // @ts-ignore Allow null
        expect(instance.transformStyles(['foo', null], { dir })).toBe('foo');
        expect(getFlushedStyles()).toMatchSnapshot();
      });

      it('flushes and purges styles from the DOM', () => {
        const styles = { test: { display: 'block' } };

        renderAndExpect(instance, styles, styles, { dir });

        instance.purgeStyles();

        expect(getFlushedStyles()).toMatchSnapshot();
      });

      describe('global sheet', () => {
        it('handles globals', () => {
          renderAndExpect(
            instance,
            SYNTAX_GLOBAL,
            {
              '@global': {
                body: { margin: 0 },
                html: { height: '100%' },
                a: {
                  color: 'red',
                  '&:hover': {
                    color: 'darkred',
                  },
                },
              },
            },
            { dir, global: true },
          );
        });

        it('handles @font-face', () => {
          renderAndExpect(
            instance,
            SYNTAX_FONT_FACE,
            {
              '@font-face': [FONT_ROBOTO_FLAT_SRC],
            },
            { dir, global: true },
          );
        });

        it('handles mixed @font-face', () => {
          renderAndExpect(
            instance,
            SYNTAX_FONT_FACE_MIXED,
            {
              '@font-face': [FONT_ROBOTO_FLAT_SRC, ...FONT_CIRCULAR_MULTIPLE_FLAT_SRC],
            },
            { dir, global: true },
          );
        });

        it('handles multiple @font-face', () => {
          renderAndExpect(
            instance,
            SYNTAX_FONT_FACE_MULTIPLE,
            {
              '@font-face': FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
            },
            { dir, global: true },
          );
        });

        it('handles @keyframes', () => {
          renderAndExpect(
            instance,
            SYNTAX_KEYFRAMES,
            {
              '@keyframes fade-0': KEYFRAME_FADE,
            },
            { dir, global: true },
          );
        });

        it('handles @viewport', () => {
          renderAndExpect(
            instance,
            SYNTAX_VIEWPORT,
            {
              '@viewport': {
                width: 'device-width',
                orientation: 'landscape',
              },
            },
            { dir, global: true },
          );
        });

        it('handles @charset', () => {
          renderAndExpect(
            instance,
            SYNTAX_CHARSET,
            {
              '@charset': '"utf8"',
            },
            { dir, global: true },
          );
        });

        it('handles single @import', () => {
          renderAndExpect(
            instance,
            SYNTAX_IMPORT,
            {
              '@import': ['url("./some/path.css")'],
            },
            { dir, global: true },
          );
        });

        it('handles multiple @import', () => {
          renderAndExpect(
            instance,
            SYNTAX_IMPORT_MULTIPLE,
            {
              '@import': ['url("./some/path.css")', 'url("./another/path.css")'],
            },
            { dir, global: true },
          );
        });
      });

      describe('style sheet', () => {
        it('converts unified syntax to native syntax and transforms to a class name', () => {
          renderAndExpect(
            instance,
            SYNTAX_UNIFIED_LOCAL_FULL,
            {
              button: {
                margin: 0,
                padding: '6px 12px',
                border: '1px solid #2e6da4',
                borderRadius: 4,
                display: 'inline-block',
                cursor: 'pointer',
                fontFamily: 'Roboto',
                fontWeight: 'normal',
                lineHeight: 'normal',
                whiteSpace: 'nowrap',
                textDecoration: 'none',
                textAlign: 'left',
                backgroundColor: '#337ab7',
                verticalAlign: 'middle',
                color: 'rgba(0, 0, 0, 0)',
                animationName: 'fade',
                animationDuration: '.3s',
                '&:hover': {
                  backgroundColor: '#286090',
                  borderColor: '#204d74',
                },
                '&::before': {
                  content: '"â˜…"',
                  display: 'inline-block',
                  verticalAlign: 'middle',
                  marginRight: 5,
                },
                '@media (max-width: 600px)': {
                  padding: '4px 8px',
                },
              },
            },
            { dir },
          );
        });

        it('handles properties', () => {
          renderAndExpect(instance, SYNTAX_PROPERTIES, SYNTAX_PROPERTIES, { dir });
        });

        it('handles attribute selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_ATTRIBUTE,
            {
              attr: {
                display: 'block',
                '&[disabled]': {
                  opacity: 0.5,
                },
              },
            },
            { dir },
          );
        });

        it('handles descendant selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_DESCENDANT,
            {
              list: {
                margin: 0,
                padding: 0,
                '&> li': {
                  listStyle: 'bullet',
                },
              },
            },
            { dir },
          );
        });

        it('handles pseudo selectors', () => {
          renderAndExpect(
            instance,
            SYNTAX_PSEUDO,
            {
              pseudo: {
                position: 'fixed',
                '&:hover': {
                  position: 'static',
                },
                '&::before': {
                  position: 'absolute',
                },
              },
            },
            { dir },
          );
        });

        it('handles multiple selectors (comma separated)', () => {
          renderAndExpect(
            instance,
            SYNTAX_MULTI_SELECTOR,
            {
              multi: {
                cursor: 'pointer',
                '&:disabled': { cursor: 'default' },
                '&[disabled]': { cursor: 'default' },
                '&> span': { cursor: 'default' },
              },
            },
            { dir },
          );
        });

        it('handles inline @keyframes', () => {
          renderAndExpect(
            instance,
            SYNTAX_KEYFRAMES_INLINE,
            {
              '@keyframes slide-0': KEYFRAME_SLIDE_PERCENT,
              '@keyframes keyframe-1-1': KEYFRAME_FADE,
              single: {
                animationName: 'slide-0',
              },
              multiple: {
                animationName: 'slide-0, unknown, keyframe-1-1',
              },
            },
            { dir },
          );
        });

        it('handles inline @font-face', () => {
          renderAndExpect(
            instance,
            SYNTAX_FONT_FACES_INLINE,
            {
              '@font-face': [
                FONT_ROBOTO_FLAT_SRC,
                ...FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
                FONT_ROBOTO_FLAT_SRC,
              ],
              single: {
                fontFamily: 'Roboto',
              },
              multiple: {
                fontFamily: 'Circular, OtherFont, Roboto',
              },
            },
            { dir },
          );
        });

        it('handles @fallbacks', () => {
          renderAndExpect(
            instance,
            SYNTAX_FALLBACKS,
            {
              fallback: {
                background: 'linear-gradient(...)',
                display: 'flex',
                fallbacks: [
                  { background: 'red' },
                  { display: 'block' },
                  { display: 'inline-block' },
                  { color: 'blue' },
                ],
              },
            },
            { dir },
          );
        });

        it('handles @media', () => {
          renderAndExpect(
            instance,
            SYNTAX_MEDIA_QUERY,
            {
              media: {
                color: 'red',
                paddingLeft: 10,
                '@media (max-width: 1000px)': {
                  color: 'green',
                  paddingLeft: 20,
                },
                '@media (min-width: 300px)': {
                  color: 'blue',
                  paddingLeft: 15,
                },
              },
            },
            { dir },
          );
        });

        it('handles nested @media', () => {
          renderAndExpect(
            instance,
            SYNTAX_MEDIA_QUERY_NESTED,
            {
              media: {
                color: 'red',
                '@media (min-width: 300px)': {
                  color: 'blue',
                  '@media (max-width: 1000px)': {
                    color: 'green',
                  },
                },
              },
            },
            { dir },
          );
        });

        it('handles @supports', () => {
          renderAndExpect(
            instance,
            SYNTAX_SUPPORTS,
            {
              sup: {
                display: 'block',
                '@supports (display: flex)': {
                  display: 'flex',
                },
                '@supports not (display: flex)': {
                  float: 'left',
                },
              },
            },
            { dir },
          );
        });

        it('handles raw CSS', () => {
          renderAndExpect(instance, SYNTAX_RAW_CSS, {}, { dir });
        });
      });
    });
  });
});
