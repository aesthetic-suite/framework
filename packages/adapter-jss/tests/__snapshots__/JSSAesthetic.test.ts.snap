// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSSAesthetic converts and transforms inline styles 1`] = `".uuid-inline-0-0-1-1 {margin: 0;}.uuid-inline-1-0-1-2 {padding: 2px;}"`;

exports[`JSSAesthetic converts and transforms inline styles 2`] = `".uuid-inline-0-0-1-1 {margin: 0;}.uuid-inline-1-0-1-2 {padding: 2px;}"`;

exports[`JSSAesthetic global sheet handles @charset 1`] = `""`;

exports[`JSSAesthetic global sheet handles @charset 2`] = `"@charset utf8;"`;

exports[`JSSAesthetic global sheet handles @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles @font-face 2`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}"`;

exports[`JSSAesthetic global sheet handles @keyframes 1`] = `""`;

exports[`JSSAesthetic global sheet handles @keyframes 2`] = `"@keyframes fade-0 {from {opacity: 0;}to {opacity: 1;}}"`;

exports[`JSSAesthetic global sheet handles @viewport 1`] = `""`;

exports[`JSSAesthetic global sheet handles @viewport 2`] = `"@viewport {width: device-width;orientation: landscape;}"`;

exports[`JSSAesthetic global sheet handles globals 1`] = `""`;

exports[`JSSAesthetic global sheet handles globals 2`] = `"body {margin: 0;}html {height: 100%;}a {color: red;}a:hover {color: darkred;}"`;

exports[`JSSAesthetic global sheet handles mixed @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles mixed @font-face 2`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}@font-face {font-family: Circular;font-style: normal;font-weight: normal;src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular;font-style: italic;font-weight: normal;src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 300;src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 700;src: url('fonts/Circular-Bold.woff2') format('woff2');}"`;

exports[`JSSAesthetic global sheet handles multiple @font-face 1`] = `""`;

exports[`JSSAesthetic global sheet handles multiple @font-face 2`] = `"@font-face {font-family: Circular;font-style: normal;font-weight: normal;src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular;font-style: italic;font-weight: normal;src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 300;src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 700;src: url('fonts/Circular-Bold.woff2') format('woff2');}"`;

exports[`JSSAesthetic global sheet handles multiple @import 1`] = `""`;

exports[`JSSAesthetic global sheet handles multiple @import 2`] = `"@import ./some/path.css;@import ./another/path.css;"`;

exports[`JSSAesthetic global sheet handles single @import 1`] = `""`;

exports[`JSSAesthetic global sheet handles single @import 2`] = `"@import ./some/path.css;"`;

exports[`JSSAesthetic style sheet converts unified syntax to native syntax and transforms to a class name 1`] = `"jss-button-0-11-1"`;

exports[`JSSAesthetic style sheet converts unified syntax to native syntax and transforms to a class name 2`] = `".jss-button-0-11-1 {color: rgba(0, 0, 0, 0);margin: 0;border: 1px solid #2e6da4;cursor: pointer;display: inline-block;padding: 6px 12px;text-align: center;font-weight: normal;line-height: normal;white-space: nowrap;font-family: Roboto;border-radius: 4px;vertical-align: middle;animation-name: fade;text-decoration: none;background-color: #337ab7;animation-duration: .3s;}.jss-button-0-11-1:hover {border-color: #204d74;background-color: #286090;}.jss-button-0-11-1::before {content: \\"â˜…\\";display: inline-block;margin-right: 5px;vertical-align: middle;}@media (max-width: 600px) {.jss-button-0-11-1 {padding: 4px 8px;}}"`;

exports[`JSSAesthetic style sheet handles @fallbacks 1`] = `"jss-fallback-0-19-1"`;

exports[`JSSAesthetic style sheet handles @fallbacks 2`] = `".jss-fallback-0-19-1 {background: red;display: block;display: inline-block;color: blue;display: flex;background: linear-gradient(...);}"`;

exports[`JSSAesthetic style sheet handles @media 1`] = `"jss-media-0-20-1"`;

exports[`JSSAesthetic style sheet handles @media 2`] = `".jss-media-0-20-1 {color: red;}@media (min-width: 300px) {.jss-media-0-20-1 {color: blue;}}@media (max-width: 1000px) {.jss-media-0-20-1 {color: green;}}"`;

exports[`JSSAesthetic style sheet handles @supports 1`] = `"jss-sup-0-22-1"`;

exports[`JSSAesthetic style sheet handles @supports 2`] = `".jss-sup-0-22-1 {display: block;}@supports (display: flex) {.jss-sup-0-22-1 {display: flex;}}@supports not (display: flex) {.jss-sup-0-22-1 {float: left;}}"`;

exports[`JSSAesthetic style sheet handles attribute selectors 1`] = `"jss-attr-0-13-1"`;

exports[`JSSAesthetic style sheet handles attribute selectors 2`] = `".jss-attr-0-13-1 {display: block;}.jss-attr-0-13-1[disabled] {opacity: 0.5;}"`;

exports[`JSSAesthetic style sheet handles descendant selectors 1`] = `"jss-list-0-14-1"`;

exports[`JSSAesthetic style sheet handles descendant selectors 2`] = `".jss-list-0-14-1 {margin: 0;padding: 0;}.jss-list-0-14-1> li {list-style: bullet;}"`;

exports[`JSSAesthetic style sheet handles inline @font-face 1`] = `"jss-single-0-18-1 jss-multiple-0-18-2"`;

exports[`JSSAesthetic style sheet handles inline @font-face 2`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}@font-face {font-family: Circular;font-style: normal;font-weight: normal;src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular;font-style: italic;font-weight: normal;src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 300;src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular;font-style: normal;font-weight: 700;src: url('fonts/Circular-Bold.woff2') format('woff2');}@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}.jss-single-0-18-1 {font-family: Roboto;}.jss-multiple-0-18-2 {font-family: Circular, OtherFont, Roboto;}"`;

exports[`JSSAesthetic style sheet handles inline @keyframes 1`] = `"jss-single-0-17-1 jss-multiple-0-17-2"`;

exports[`JSSAesthetic style sheet handles inline @keyframes 2`] = `"@keyframes slide-0 {0% {opacity: 0;}50% {opacity: 0.5;}100% {opacity: 1;}}@keyframes keyframe-1-1 {from {opacity: 0;}to {opacity: 1;}}.jss-single-0-17-1 {animation-name: slide-0;}.jss-multiple-0-17-2 {animation-name: slide-0, unknown, keyframe-1-1;}"`;

exports[`JSSAesthetic style sheet handles multiple selectors (comma separated) 1`] = `"jss-multi-0-16-1"`;

exports[`JSSAesthetic style sheet handles multiple selectors (comma separated) 2`] = `".jss-multi-0-16-1 {cursor: pointer;}.jss-multi-0-16-1:disabled {cursor: default;}.jss-multi-0-16-1[disabled] {cursor: default;}.jss-multi-0-16-1> span {cursor: default;}"`;

exports[`JSSAesthetic style sheet handles nested @media 1`] = `"jss-media-0-21-1"`;

exports[`JSSAesthetic style sheet handles nested @media 2`] = `".jss-media-0-21-1 {color: red;}@media (min-width: 300px) {.jss-media-0-21-1 {color: blue;}@media (max-width: 1000px) {.jss-media-0-21-1 {color: green;}}}"`;

exports[`JSSAesthetic style sheet handles properties 1`] = `"jss-props-0-12-1"`;

exports[`JSSAesthetic style sheet handles properties 2`] = `".jss-props-0-12-1 {color: black;margin: 10px;display: inline;}"`;

exports[`JSSAesthetic style sheet handles pseudo selectors 1`] = `"jss-pseudo-0-15-1"`;

exports[`JSSAesthetic style sheet handles pseudo selectors 2`] = `".jss-pseudo-0-15-1 {position: fixed;}.jss-pseudo-0-15-1:hover {position: static;}.jss-pseudo-0-15-1::before {position: absolute;}"`;

exports[`JSSAesthetic style sheet handles raw CSS 1`] = `""`;

exports[`JSSAesthetic style sheet handles raw CSS 2`] = `".jss-button{display:'block';font-size:16px;}.jss-button:hover{color:'red';}.jss-button{vertical-align:'middle';}@media (max-width:600px){.jss-button{display:'none';}}@supports (display:-webkit-box) or (display:-webkit-flex) or (display:-ms-flexbox) or (display:flex){.jss-button{display:'flex';}}"`;
